components/sup/src/ctl_gateway/mod.rs

```diff
@@ -163,56 +160,65 @@ impl Write for CtlRequest {
     fn flush(&mut self) -> io::Result<()> { Ok(()) }
 
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let line = String::from_utf8(buf.to_vec()).expect("CtlRequest buffer valid utf8");
+
         // The protocol reply is destined for the client, so (for now,
         // at least), we'll apply colored output.
         //
         // `line` will also have a newline character at the end, FYI.
         let mut msg = protocol::ctl::ConsoleLine::default();
-        msg.line = String::from_utf8_lossy(buf).to_string();
-        match self.current_color_spec {
-            Some(ref spec) => {
-                msg.color = color_to_string(spec.fg());
-                msg.bold = spec.bold();
-            }
-            None => {
-                msg.color = None;
-                msg.bold = false;
-            }
-        }
+        msg.line = line.to_string();
+        msg.color = color_to_string(self.current_color_spec.fg());
+        msg.bold = self.current_color_spec.bold();
         self.reply_partial(msg);
 
         // Down here, however, we're doing double-duty by *also*
-        // sending the output to the Supervisor's output stream. In
-        // _this_ case, we want to honor whatever global logging flags
-        // have been set; in particular, if we're emitting
-        // JSON-formatted logs or if we've been instructed to not
-        // output ANSI color codes, we need to strip those codes out.
+        // sending the output to the Supervisor's output stream.
         //
         // TODO (CM): Obviously, this feels hacky to the extreme. It'd
         // be nice to find a cleaner way to model this, but the fact
         // that CtlRequest is sending output to two destinations with
         // different formatting requirements complicates things a bit.
-        //
-        // TODO (MW): For `hab sup run` scenarios, it would be nice to
-        // support color on older versions of Windows. We could refactor
-        // outputln! to support a cross platform ColorWriter similar to
-        // what we use in common::UI.
-        let mut ansi_buffer = Buffer::ansi();
-        if let Some(spec) = &self.current_color_spec {
-            ansi_buffer.set_color(&spec)?;
-        }
-        ansi_buffer.write_all(buf)?;
-        ansi_buffer.flush()?;
-        let line = String::from_utf8_lossy(ansi_buffer.as_slice()).to_string();
 
-        let maybe_stripped = if output::is_json() || !output::is_color() {
-            STRIP_ANSI_CODES.replace_all(&line, "")
+        let is_line_ending = line.ends_with('\n');
+        if self.is_new_line || output::get_format() == OutputFormat::JSON {
```

Comment:
This is a somewhat surprising / confusing if condition... why is "newline + JSON" the criterion?

Reply:
yeah this could use some comments that I will add. Previous to 0.77, `write` on the `CtlRequest` would be called for each line of output. Now it can be called for line fragments when styling changes mid-line. Unless we are rendering JSON, we only want a `StructuredOutput` line header when we are beginning a new line, otherwise we just want the message content.